File: port_killer.py
#!/usr/bin/env python3
"""
╔═══════════════════════════════════════════════════════╗
║         PORT SCANNER & KILLER  —  Day 6/30            ║
║         30-Day GitHub Roadmap                         ║
╚═══════════════════════════════════════════════════════╝

Find which process is hogging a port and kill it in one command.

Usage:
  python port_killer.py scan                  # Scan common ports
  python port_killer.py scan --range 1-1024   # Scan port range
  python port_killer.py find 3000             # Find process on port 3000
  python port_killer.py kill 3000             # Kill process on port 3000
  python port_killer.py kill 3000 8080 5432   # Kill multiple ports
  python port_killer.py watch 3000            # Watch a port (live)
"""

import socket
import sys
import os
import argparse
import time
import platform
from concurrent.futures import ThreadPoolExecutor, as_completed

# ── Colour helpers (no deps) ──────────────────────────────────────────────────
RESET  = "\033[0m"
BOLD   = "\033[1m"
DIM    = "\033[2m"
RED    = "\033[91m"
GREEN  = "\033[92m"
YELLOW = "\033[93m"
CYAN   = "\033[96m"
WHITE  = "\033[97m"
GRAY   = "\033[90m"

IS_WINDOWS = platform.system() == "Windows"

def c(text, *codes): return "".join(codes) + str(text) + RESET
def ok(msg):   print(c("  ✔ ", GREEN, BOLD) + msg)
def err(msg):  print(c("  ✘ ", RED,   BOLD) + msg)
def info(msg): print(c("  ℹ ", CYAN,  BOLD) + msg)
def warn(msg): print(c("  ⚠ ", YELLOW,BOLD) + msg)

# ── Common ports dictionary ───────────────────────────────────────────────────
KNOWN_PORTS = {
    20: "FTP Data",       21: "FTP Control",    22: "SSH",
    23: "Telnet",         25: "SMTP",            53: "DNS",
    80: "HTTP",           110: "POP3",           143: "IMAP",
    443: "HTTPS",         465: "SMTPS",          587: "SMTP TLS",
    993: "IMAPS",         995: "POP3S",          1433: "MSSQL",
    1521: "Oracle DB",    2181: "Zookeeper",     2375: "Docker",
    3000: "Dev Server",   3001: "Dev Server Alt",3306: "MySQL",
    3389: "RDP",          4200: "Angular Dev",   4443: "HTTPS Alt",
    5000: "Flask/Dev",    5173: "Vite Dev",      5432: "PostgreSQL",
    5900: "VNC",          6379: "Redis",         6443: "Kubernetes",
    7474: "Neo4j",        8000: "HTTP Alt",      8080: "HTTP Proxy",
    8081: "HTTP Alt 2",   8443: "HTTPS Alt",     8888: "Jupyter",
    9000: "Portainer",    9090: "Prometheus",    9200: "Elasticsearch",
    9300: "ES Transport", 27017: "MongoDB",      27018: "MongoDB Alt",
}

# ── Port probing ──────────────────────────────────────────────────────────────
def is_port_open(port: int, host: str = "127.0.0.1", timeout: float = 0.5) -> bool:
    try:
        with socket.create_connection((host, port), timeout=timeout):
            return True
    except (ConnectionRefusedError, socket.timeout, OSError):
        return False

def get_process_on_port(port: int):
    """Returns (pid, name, command) or None. Uses psutil if available, else falls back to OS commands."""
    try:
        import psutil
        for conn in psutil.net_connections(kind="inet"):
            if conn.laddr.port == port and conn.status in ("LISTEN", "ESTABLISHED"):
                try:
                    proc = psutil.Process(conn.pid)
                    return conn.pid, proc.name(), " ".join(proc.cmdline())
                except (psutil.NoSuchProcess, psutil.AccessDenied):
                    return conn.pid, "unknown", ""
        return None
    except ImportError:
        return _get_process_fallback(port)

def _get_process_fallback(port: int):
    """Fallback using system commands when psutil is not installed."""
    import subprocess
    try:
        if IS_WINDOWS:
            cmd = f'netstat -ano | findstr ":{port} "'
            result = subprocess.check_output(cmd, shell=True, text=True, stderr=subprocess.DEVNULL)
            for line in result.strip().splitlines():
                parts = line.split()
                if len(parts) >= 5:
                    pid = int(parts[-1])
                    name_cmd = f'tasklist /FI "PID eq {pid}" /FO CSV /NH'
                    name_res = subprocess.check_output(name_cmd, shell=True, text=True, stderr=subprocess.DEVNULL)
                    name = name_res.strip().split(",")[0].strip('"') if name_res.strip() else "unknown"
                    return pid, name, ""
        else:
            # lsof approach
            result = subprocess.check_output(
                ["lsof", "-i", f":{port}", "-sTCP:LISTEN", "-n", "-P"],
                text=True, stderr=subprocess.DEVNULL
            )
            lines = result.strip().splitlines()
            if len(lines) > 1:
                parts = lines[1].split()
                name = parts[0]
                pid  = int(parts[1])
                return pid, name, " ".join(parts[8:]) if len(parts) > 8 else ""
    except Exception:
        pass
    return None

def kill_process(pid: int, force: bool = False) -> bool:
    try:
        import psutil
        proc = psutil.Process(pid)
        if force:
            proc.kill()
        else:
            proc.terminate()
            proc.wait(timeout=3)
        return True
    except ImportError:
        return _kill_fallback(pid, force)
    except Exception as e:
        err(f"Failed to kill PID {pid}: {e}")
        return False

def _kill_fallback(pid: int, force: bool = False) -> bool:
    import subprocess
    try:
        if IS_WINDOWS:
            subprocess.run(["taskkill", "/PID", str(pid), "/F"], check=True, capture_output=True)
        else:
            sig = "-9" if force else "-15"
            subprocess.run(["kill", sig, str(pid)], check=True, capture_output=True)
        return True
    except Exception as e:
        err(f"Failed to kill PID {pid}: {e}")
        return False

# ── Banner ────────────────────────────────────────────────────────────────────
def banner():
    print(c("""
 ██████╗  ██████╗ ██████╗ ████████╗    ██╗  ██╗██╗██╗     ██╗     ███████╗██████╗
 ██╔══██╗██╔═══██╗██╔══██╗╚══██╔══╝    ██║ ██╔╝██║██║     ██║     ██╔════╝██╔══██╗
 ██████╔╝██║   ██║██████╔╝   ██║       █████╔╝ ██║██║     ██║     █████╗  ██████╔╝
 ██╔═══╝ ██║   ██║██╔══██╗   ██║       ██╔═██╗ ██║██║     ██║     ██╔══╝  ██╔══██╗
 ██║     ╚██████╔╝██║  ██║   ██║       ██║  ██╗██║███████╗███████╗███████╗██║  ██║
 ╚═╝      ╚═════╝ ╚═╝  ╚═╝   ╚═╝       ╚═╝  ╚═╝╚═╝╚══════╝╚══════╝╚══════╝╚═╝  ╚═╝
""", CYAN, BOLD))
    print(c("  Day 06/30 · 30-Day GitHub Roadmap", GRAY))
    print(c("  Find & kill processes hogging your ports.\n", GRAY))

# ── Commands ──────────────────────────────────────────────────────────────────
def cmd_scan(port_range: str = None, host: str = "127.0.0.1"):
    """Scan ports for open connections."""
    if port_range:
        try:
            start, end = map(int, port_range.split("-"))
            ports = list(range(start, end + 1))
            info(f"Scanning ports {start}–{end} on {host} ...")
        except ValueError:
            err("Invalid range format. Use: --range 1-1024")
            sys.exit(1)
    else:
        ports = list(KNOWN_PORTS.keys())
        info(f"Scanning {len(ports)} common ports on {host} ...")

    print()
    open_ports = []

    with ThreadPoolExecutor(max_workers=100) as ex:
        futures = {ex.submit(is_port_open, p, host): p for p in ports}
        for future in as_completed(futures):
            port = futures[future]
            if future.result():
                open_ports.append(port)

    if not open_ports:
        warn("No open ports found.")
        return

    open_ports.sort()
    print(f"  {'PORT':<8} {'SERVICE':<22} {'PID':<8} {'PROCESS'}")
    print(f"  {'─'*8} {'─'*22} {'─'*8} {'─'*28}")

    for port in open_ports:
        service = KNOWN_PORTS.get(port, "Unknown")
        proc    = get_process_on_port(port)
        if proc:
            pid, name, _ = proc
            print(f"  {c(port, GREEN, BOLD):<17} {c(service, CYAN):<31} {c(pid, YELLOW):<17} {name}")
        else:
            print(f"  {c(port, GREEN, BOLD):<17} {c(service, CYAN):<31} {'—':<8} {'—'}")

    print(f"\n  {c(len(open_ports), GREEN, BOLD)} open port(s) found.\n")


def cmd_find(port: int, host: str = "127.0.0.1"):
    """Find the process using a specific port."""
    info(f"Looking up port {c(port, CYAN, BOLD)} on {host} ...\n")

    if not is_port_open(port, host):
        warn(f"Port {port} appears to be closed or not listening.")
        return

    proc = get_process_on_port(port)
    if not proc:
        warn(f"Port {port} is open but couldn't identify the process.")
        warn("Try running with sudo/administrator privileges.")
        return

    pid, name, cmd = proc
    print(f"  {c('PORT', BOLD)}     {c(port, GREEN, BOLD)}")
    print(f"  {c('PID', BOLD)}      {c(pid, YELLOW, BOLD)}")
    print(f"  {c('PROCESS', BOLD)}  {c(name, CYAN)}")
    if cmd:
        print(f"  {c('COMMAND', BOLD)}  {c(cmd[:80] + ('...' if len(cmd) > 80 else ''), GRAY)}")
    print()
    info(f"To kill it: {c(f'python port_killer.py kill {port}', WHITE)}\n")


def cmd_kill(ports: list, host: str = "127.0.0.1", force: bool = False):
    """Kill processes on given ports."""
    for port in ports:
        print()
        info(f"Targeting port {c(port, CYAN, BOLD)} ...")

        if not is_port_open(port, host):
            warn(f"Port {port} is not open. Nothing to kill.")
            continue

        proc = get_process_on_port(port)
        if not proc:
            warn(f"Couldn't identify process on port {port}. Try with sudo.")
            continue

        pid, name, cmd = proc
        print(f"  {c('Found:', BOLD)} {c(name, CYAN)} (PID {c(pid, YELLOW)})")

        confirm = input(f"  {c('Kill this process?', WHITE)} {c('[y/N]', GRAY)} ").strip().lower()
        if confirm not in ("y", "yes"):
            warn("Skipped.")
            continue

        if kill_process(pid, force=force):
            ok(f"Process {c(name, CYAN)} (PID {c(pid, YELLOW)}) terminated.")
            # verify
            time.sleep(0.5)
            if not is_port_open(port, host):
                ok(f"Port {c(port, GREEN)} is now free! 🎉")
            else:
                warn(f"Port {port} still appears open. Try with --force.")
        else:
            err(f"Failed to kill PID {pid}.")

    print()


def cmd_watch(port: int, host: str = "127.0.0.1", interval: float = 2.0):
    """Live-watch a port and show status changes."""
    info(f"Watching port {c(port, CYAN, BOLD)} every {interval}s. Press Ctrl+C to stop.\n")
    last_state = None

    try:
        while True:
            open_now = is_port_open(port, host)
            ts = time.strftime("%H:%M:%S")

            if open_now != last_state:
                if open_now:
                    proc = get_process_on_port(port)
                    detail = f"(PID {proc[1]})" if proc else ""
                    print(f"  {c(ts, GRAY)}  {c('OPEN  ▲', GREEN, BOLD)}  {detail}")
                else:
                    print(f"  {c(ts, GRAY)}  {c('CLOSED ▼', RED, BOLD)}")
                last_state = open_now
            else:
                state_str = c("OPEN  ●", GREEN) if open_now else c("CLOSED ●", RED)
                print(f"  {c(ts, GRAY)}  {state_str}", end="\r")

            time.sleep(interval)

    except KeyboardInterrupt:
        print(f"\n\n  {c('Watch stopped.', GRAY)}\n")


# ── Entry point ───────────────────────────────────────────────────────────────
def main():
    banner()

    parser = argparse.ArgumentParser(
        description="Port Scanner & Killer — find and kill processes hogging ports.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python port_killer.py scan
  python port_killer.py scan --range 3000-9000
  python port_killer.py find 8080
  python port_killer.py kill 3000
  python port_killer.py kill 3000 8080 5432
  python port_killer.py watch 5000
        """
    )

    sub = parser.add_subparsers(dest="command", required=True)

    # scan
    p_scan = sub.add_parser("scan", help="Scan common (or ranged) ports")
    p_scan.add_argument("--range", dest="range", help="Port range e.g. 1-1024")
    p_scan.add_argument("--host", default="127.0.0.1", help="Host to scan (default: 127.0.0.1)")

    # find
    p_find = sub.add_parser("find", help="Find process on a specific port")
    p_find.add_argument("port", type=int, help="Port number")
    p_find.add_argument("--host", default="127.0.0.1")

    # kill
    p_kill = sub.add_parser("kill", help="Kill process(es) on port(s)")
    p_kill.add_argument("ports", type=int, nargs="+", help="One or more port numbers")
    p_kill.add_argument("--host", default="127.0.0.1")
    p_kill.add_argument("--force", action="store_true", help="Force kill (SIGKILL)")

    # watch
    p_watch = sub.add_parser("watch", help="Live-watch a port for state changes")
    p_watch.add_argument("port", type=int, help="Port number")
    p_watch.add_argument("--host", default="127.0.0.1")
    p_watch.add_argument("--interval", type=float, default=2.0, help="Poll interval seconds (default: 2)")

    args = parser.parse_args()

    if args.command == "scan":
        cmd_scan(port_range=args.range, host=args.host)
    elif args.command == "find":
        cmd_find(port=args.port, host=args.host)
    elif args.command == "kill":
        cmd_kill(ports=args.ports, host=args.host, force=args.force)
    elif args.command == "watch":
        cmd_watch(port=args.port, host=args.host, interval=args.interval)


if __name__ == "__main__":
    main()

File: port_killer.js
// ── Data ──────────────────────────────────────────────────────────────────
  const KNOWN_PORTS = {
    20:"FTP Data",21:"FTP Control",22:"SSH",23:"Telnet",25:"SMTP",
    53:"DNS",80:"HTTP",110:"POP3",143:"IMAP",443:"HTTPS",
    465:"SMTPS",587:"SMTP TLS",993:"IMAPS",995:"POP3S",1433:"MSSQL",
    1521:"Oracle DB",2181:"Zookeeper",2375:"Docker",3000:"Dev Server",
    3001:"Dev Server Alt",3306:"MySQL",3389:"RDP",4200:"Angular Dev",
    4443:"HTTPS Alt",5000:"Flask/Dev",5173:"Vite Dev",5432:"PostgreSQL",
    5900:"VNC",6379:"Redis",6443:"Kubernetes",7474:"Neo4j",
    8000:"HTTP Alt",8080:"HTTP Proxy",8081:"HTTP Alt 2",8443:"HTTPS Alt",
    8888:"Jupyter",9000:"Portainer",9090:"Prometheus",9200:"Elasticsearch",
    9300:"ES Transport",27017:"MongoDB",27018:"MongoDB Alt"
  };

  // Simulated open ports for demo purposes
  const DEMO_OPEN = {
    3000: { pid: 48231, name: "node",          cmd: "node server.js" },
    5432: { pid: 1042,  name: "postgres",      cmd: "postgres -D /usr/local/var/postgres" },
    6379: { pid: 892,   name: "redis-server",  cmd: "redis-server 127.0.0.1:6379" },
    8080: { pid: 50341, name: "python",        cmd: "python -m http.server 8080" },
    27017:{ pid: 2203,  name: "mongod",        cmd: "mongod --config /etc/mongod.conf" },
    5173: { pid: 61022, name: "node (vite)",   cmd: "vite --port 5173" },
  };

  // Render the ports reference on the about page
  document.getElementById('tab-about').innerHTML = document.getElementById('tab-about').innerHTML.replace(
    '${Object.entries(KNOWN_PORTS_HTML).map(([p,s]) => `<div style="display:flex;gap:8px;align-items:center;padding:5px 8px;background:var(--card);border-radius:4px;border:1px solid var(--border)"><span style="color:var(--green);font-weight:700;min-width:40px">${p}</span><span style="color:var(--muted)">${s}</span></div>`).join(\'\')}',
    Object.entries(KNOWN_PORTS).map(([p,s]) =>
      `<div style="display:flex;gap:8px;align-items:center;padding:5px 8px;background:var(--card);border-radius:4px;border:1px solid var(--border)">
        <span style="color:var(--green);font-weight:700;min-width:40px">${p}</span>
        <span style="color:var(--muted)">${s}</span>
       </div>`
    ).join('')
  );

  // ── Tab switching ─────────────────────────────────────────────────────────
  function switchTab(name, btn) {
    document.querySelectorAll('.tab-pane').forEach(t => t.classList.remove('active'));
    document.querySelectorAll('.nav-btn').forEach(b => b.classList.remove('active'));
    document.getElementById('tab-' + name).classList.add('active');
    btn.classList.add('active');
    if (name !== 'watch') stopWatch();
  }

  // ── Terminal log ──────────────────────────────────────────────────────────
  function addLog(el, text, cls = '') {
    const line = document.createElement('div');
    line.className = 'log-line ' + cls;
    line.textContent = text;
    el.appendChild(line);
    el.scrollTop = el.scrollHeight;
  }

  // ── SCAN ──────────────────────────────────────────────────────────────────
  let scanTimer = null;

  function setScanState(state, status, count = '') {
    const dot  = document.getElementById('scanDot');
    const txt  = document.getElementById('scanStatus');
    const cnt  = document.getElementById('scanCount');
    dot.className = 'status-dot ' + state;
    txt.textContent = status;
    cnt.textContent = count;
  }

  function clearResults() {
    document.getElementById('resultsBody').innerHTML = `
      <div class="empty-state">
        <div class="empty-icon">🔌</div>
        <div class="empty-text">No scan results yet.<br>Run a scan to see open ports.</div>
      </div>`;
    setScanState('', 'Ready. Select a range and run scan.', '');
    const log = document.getElementById('termLog');
    log.innerHTML = '';
    addLog(log, '$ port_killer.py scan', 'log-cmd');
    addLog(log, 'Results cleared.', 'log-warn');
  }

  function quickScan() {
    document.getElementById('rangeStart').value = 1;
    document.getElementById('rangeEnd').value = 9999;
    runScan();
  }

  function runScan() {
    const start = parseInt(document.getElementById('rangeStart').value);
    const end   = parseInt(document.getElementById('rangeEnd').value);
    const host  = document.getElementById('hostInput').value || '127.0.0.1';
    const log   = document.getElementById('termLog');
    const body  = document.getElementById('resultsBody');

    if (start > end) { setScanState('error', 'Error: start must be ≤ end.'); return; }

    log.innerHTML = '';
    addLog(log, `$ port_killer.py scan --range ${start}-${end} --host ${host}`, 'log-cmd');
    addLog(log, `[INFO] Scanning ${end - start + 1} ports on ${host}...`, 'log-info');

    body.innerHTML = '';
    setScanState('scanning', `Scanning ports ${start}–${end} on ${host}…`);

    let found = 0;
    const openPorts = Object.keys(DEMO_OPEN)
      .map(Number)
      .filter(p => p >= start && p <= end);

    let i = 0;
    const total = openPorts.length;

    function step() {
      if (i >= total) {
        setScanState('done', `Scan complete.`, `${found} open`);
        addLog(log, `[OK] Scan finished. ${found} open port(s) found.`, 'log-ok');
        if (found === 0) {
          body.innerHTML = `<div class="empty-state"><div class="empty-icon">✅</div><div class="empty-text">No open ports found in range ${start}–${end}.</div></div>`;
        }
        return;
      }
      const port = openPorts[i++];
      const info = DEMO_OPEN[port];
      const service = KNOWN_PORTS[port] || 'Unknown';
      addLog(log, `[OPEN] Port ${port} → ${info.name} (PID ${info.pid})`, 'log-ok');
      found++;
      setScanState('scanning', `Scanning… found ${found} open port(s)`, `${found} open`);

      const row = document.createElement('div');
      row.className = 'result-row';
      row.style.animationDelay = (i * 60) + 'ms';
      row.innerHTML = `
        <div class="port-num">${port}</div>
        <div class="service-name">${service}</div>
        <div class="proc-name">${info.name}</div>
        <div class="pid-val">${info.pid}</div>
        <div><button class="kill-btn" onclick="quickKill(${port})" title="Kill port ${port}">✕</button></div>
      `;
      body.appendChild(row);
      body.scrollTop = body.scrollHeight;

      setTimeout(step, 180 + Math.random() * 120);
    }

    setTimeout(step, 300);
  }

  function quickKill(port) {
    if (!confirm(`Kill process on port ${port}?`)) return;
    delete DEMO_OPEN[port];
    document.querySelectorAll('.result-row').forEach(row => {
      if (row.querySelector('.port-num')?.textContent == port) {
        row.style.background = 'rgba(248,81,73,0.1)';
        row.style.opacity = '0.4';
        setTimeout(() => row.remove(), 600);
      }
    });
    const log = document.getElementById('termLog');
    addLog(log, `[OK] Process on port ${port} terminated.`, 'log-ok');
    const count = document.getElementById('scanCount');
    const n = parseInt(count.textContent) - 1;
    count.textContent = `${n} open`;
  }

  // ── FIND ──────────────────────────────────────────────────────────────────
  function runFind() {
    const port = parseInt(document.getElementById('findPort').value);
    const host = document.getElementById('hostInput').value || '127.0.0.1';
    const el   = document.getElementById('findResult');
    const info = DEMO_OPEN[port];

    if (!info) {
      el.innerHTML = `
        <div style="background:var(--panel);border:1px solid rgba(248,81,73,0.3);border-radius:8px;padding:24px 28px;color:var(--red);">
          <div style="font-size:13px;font-weight:700;margin-bottom:6px;">⚠ Port ${port} is not open</div>
          <div style="font-size:11px;color:var(--muted);">No process found listening on port ${port} at ${host}.</div>
        </div>`;
      return;
    }

    const service = KNOWN_PORTS[port] || 'Unknown';
    el.innerHTML = `
      <div class="find-result">
        <div class="find-field">
          <div class="find-label">Port</div>
          <div class="find-value" style="color:var(--green)">${port}</div>
        </div>
        <div class="find-field">
          <div class="find-label">Service</div>
          <div class="find-value" style="color:var(--cyan)">${service}</div>
        </div>
        <div class="find-field">
          <div class="find-label">PID</div>
          <div class="find-value" style="color:var(--yellow)">${info.pid}</div>
        </div>
        <div class="find-field">
          <div class="find-label">Process Name</div>
          <div class="find-value" style="color:var(--text)">${info.name}</div>
        </div>
        <div class="find-field" style="grid-column:1/-1">
          <div class="find-label">Command</div>
          <div class="find-value" style="font-size:11px;color:var(--muted);font-weight:400">${info.cmd}</div>
        </div>
        <div class="find-field" style="grid-column:1/-1">
          <div class="find-label">Kill Command</div>
          <div style="background:var(--bg);border:1px solid var(--border2);border-radius:5px;padding:8px 12px;font-size:11px;color:var(--blue)">python port_killer.py kill ${port}</div>
        </div>
      </div>`;
  }

  // ── KILLER ──────────────────────────────────────────────────────────────────
  function runKill() {
    const raw   = document.getElementById('killPorts').value;
    const ports = raw.split(',').map(s => parseInt(s.trim())).filter(n => !isNaN(n));
    const log   = document.getElementById('killLogBody');
    log.innerHTML = '';

    addLog(log, `$ port_killer.py kill ${ports.join(' ')}`, 'log-cmd');

    ports.forEach((port, idx) => {
      setTimeout(() => {
        const info = DEMO_OPEN[port];
        if (!info) {
          addLog(log, `[WARN] Port ${port}: not open or already free.`, 'log-warn');
          return;
        }
        addLog(log, `[INFO] Port ${port}: found ${info.name} (PID ${info.pid})`, 'log-info');
        setTimeout(() => {
          addLog(log, `[OK]   Port ${port}: process ${info.name} terminated. Port is now free.`, 'log-ok');
          delete DEMO_OPEN[port];
        }, 500);
      }, idx * 700);
    });
  }

  // ── WATCH ─────────────────────────────────────────────────────────────────
  let watchTimer = null;
  let isWatching = false;
  let lastWatchState = null;

  function toggleWatch() {
    if (isWatching) stopWatch();
    else startWatch();
  }

  function startWatch() {
    const port     = parseInt(document.getElementById('watchPort').value);
    const interval = parseFloat(document.getElementById('watchInterval').value) * 1000;
    const btn      = document.getElementById('watchStartBtn');
    const ring     = document.getElementById('watchRing');
    const portDisp = document.getElementById('watchPortDisplay');
    const badge    = document.getElementById('watchStatusBadge');
    const watchLog = document.getElementById('watchLog');

    isWatching = true;
    btn.textContent = '⏹ Stop Watch';
    btn.className = 'btn btn-danger';
    portDisp.textContent = port;
    portDisp.style.fontSize = '72px';
    watchLog.innerHTML = '';

    function check() {
      const isOpen = !!DEMO_OPEN[port];
      const ts = new Date().toLocaleTimeString();

      if (isOpen !== lastWatchState) {
        const line = document.createElement('div');
        line.style.cssText = `color:${isOpen ? 'var(--green)' : 'var(--red)'};`;
        line.textContent = `${ts}  ${isOpen ? '▲ OPEN' : '▼ CLOSED'}`;
        watchLog.appendChild(line);
        watchLog.scrollTop = watchLog.scrollHeight;
        lastWatchState = isOpen;
      }

      ring.className = 'watch-ring ' + (isOpen ? 'open' : 'closed');
      portDisp.style.color = isOpen ? 'var(--green)' : 'var(--red)';
      badge.style.background = isOpen ? 'rgba(61,214,140,0.1)' : 'rgba(248,81,73,0.1)';
      badge.style.color = isOpen ? 'var(--green)' : 'var(--red)';
      badge.style.border = `1px solid ${isOpen ? 'rgba(61,214,140,0.3)' : 'rgba(248,81,73,0.3)'}`;
      badge.textContent = isOpen ? 'OPEN' : 'CLOSED';
    }

    check();
    watchTimer = setInterval(check, interval);
  }

  function stopWatch() {
    clearInterval(watchTimer);
    isWatching = false;
    lastWatchState = null;
    const btn = document.getElementById('watchStartBtn');
    btn.textContent = '▶ Start Watch';
    btn.className = 'btn btn-primary';
    document.getElementById('watchRing').className = 'watch-ring';
    document.getElementById('watchPortDisplay').textContent = '—';
    document.getElementById('watchPortDisplay').style.color = 'var(--muted2)';
    document.getElementById('watchPortDisplay').style.fontSize = '36px';
    document.getElementById('watchStatusBadge').textContent = 'IDLE';
    document.getElementById('watchStatusBadge').style.cssText = 'background:var(--card);color:var(--muted);border:1px solid var(--border2)';
  }
